pmd-java/etc/grammar/Java.jjt

```diff
@@ -1073,6 +1107,67 @@ void EnumConstant():
   t=<IDENTIFIER> {jjtThis.setImage(t.image);} [ Arguments() ] [ ClassOrInterfaceBody() ]
 }
 
+void RecordDeclaration(int modifiers):
+{
+    Token t;
+    jjtThis.setModifiers(modifiers);
+    checkForRecordType();
+}
+{
+  t = <IDENTIFIER> {
+    if (!"record".equals(t.image)) {
+      throw new ParseException("ERROR: expecting record");
+    }
+  }
+  t=<IDENTIFIER> {checkForBadTypeIdentifierUsage(t.image); jjtThis.setImage(t.image);}
+  [ TypeParameters() ]
+  RecordComponentList()
+  [ ImplementsList() ]
+  RecordBody()
+}
+
+void RecordComponentList() :
+{}
+{
+  "(" [ RecordComponent() ("," RecordComponent())* ] ")"
+}
+
+void RecordComponent():
+{}
+{
+  (Annotation())*
+  Type()
+  [ "..." {jjtThis.setVarargs();} ]
+  VariableDeclaratorId()
+}
+
+void RecordBody():
+{}
+{
+  "{"
+    ( RecordBodyDeclaration() )*
+  "}"
+}
+
+void RecordBodyDeclaration() #void :
+{}
+{
+  LOOKAHEAD(ClassOrInterfaceBodyDeclaration()) ClassOrInterfaceBodyDeclaration()
+  |
+  RecordConstructorDeclaration()
+}
+
+void RecordConstructorDeclaration():
+{
+  int modifiers;
+}
+{
+  modifiers = Modifiers() { jjtThis.setModifiers(modifiers); }
+  [TypeParameters()]
+  <IDENTIFIER> { jjtThis.setImage(token.image); }
+  "{" ( BlockStatement() )* "}"
```

Comment:
Actually a record ctor may have a throws list. I'll fix that shortly.

Reply:
I think I've removed this, because the EA build didn't compile it - at least with the version I used (build 14+36-1461). The JEPs actually mentions the throws clause in the grammar, but not the JLS update - we need to see, whether other exceptions than runtime exceptions are actually allowed in a record constructor...

Reply:
Ah I see! I found, that they talked about this, and you're right: